File: app/src/main/java/com/example/amit/uniconnexample/Signupactivity.java
package com.example.amit.uniconnexample;

import android.*;
import android.Manifest;
import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PorterDuff;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import org.apache.commons.validator.routines.EmailValidator;

import java.io.File;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.internal.Utils;
import pl.aprilapps.easyphotopicker.DefaultCallback;
import pl.aprilapps.easyphotopicker.EasyImage;
import timber.log.Timber;

import static com.example.amit.uniconnexample.R.drawable.user;
import static com.example.amit.uniconnexample.R.id.photo;

/**
 * Created by amit on 29/10/16.
 */

public class Signupactivity extends AppCompatActivity {
   @BindView(R.id.name)
   TextInputLayout name;
   @BindView(R.id.email)
   TextInputLayout email;
   @BindView(R.id.password)
   TextInputLayout password;
   @BindView(R.id.phone)
   TextInputLayout phone;
    @BindView(R.id.confirmpassword)
    TextInputLayout confrmpassword;
    @BindView(R.id.iview)
    ImageView iv;
    @BindView(R.id.clg)
    TextInputLayout clg;
    @BindView(R.id.sign_up)
    Button signup;
    @BindView(R.id.person)
     ImageView person;
    @BindView(R.id.msg)
    ImageView msg;
    @BindView(R.id.lock1)
    ImageView lock1;
    @BindView(R.id.lock2)
    ImageView lock2;
    @BindView(R.id.college)
    ImageView college;
    @BindView(R.id.phn)
    ImageView phonen;
    FirebaseUser user;
    private ProgressDialog mProgress;
    private FirebaseAuth mAuth;
    private FirebaseAuth.AuthStateListener mAuthListener;
    UserData userData;
    DatabaseReference mDatabasenotiflike;
    String mal,pass,confrmpass,phn,nam,clgname;
    int flag=0;
    String check;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);
        ButterKnife.bind(this);
        mProgress = new ProgressDialog(this);
        userData = new UserData();
        mDatabasenotiflike=FirebaseDatabase.getInstance().getReference().child("notificationdata").child("like");
        phn=phone.getEditText().getText().toString().trim();
        pass=password.getEditText().getText().toString().trim();
        confrmpass=confrmpassword.getEditText().getText().toString().trim();
        nam=name.getEditText().getText().toString().trim();
        mal=email.getEditText().getText().toString().trim();
        clgname=clg.getEditText().getText().toString().trim();
        mAuth = FirebaseAuth.getInstance();
        int tabIconColor = ContextCompat.getColor(getBaseContext(), R.color.md_orange_600);
        person.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN);
        msg.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN);
        lock1.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN);
        lock2.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN);
        college.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN);
        phonen.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN);
     //    user=FirebaseAuth.getInstance().getCurrentUser();
       mAuthListener = new FirebaseAuth.AuthStateListener() {
            @Override
            public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
                 user = firebaseAuth.getCurrentUser();
                if (user != null) {                    // User is signed in
                    Toast.makeText(Signupactivity.this, "Successfully signed up", Toast.LENGTH_SHORT).show();

                    writeUserData(user.getUid());
              //      mDatabasenotiflike.child(user.getUid()).setValue(new Likemodel(0));
                    Intent intent=new Intent(Signupactivity.this, Loginactivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                    startActivity(intent);
                           // .putExtra("user", userData));
                    //finish();
                } else {                    // User is signed out
                    Timber.d("User is signed out");
                }
            }
        };
    }
    private void writeUserData(String uid) {
        FirebaseDatabase database = FirebaseDatabase.getInstance();
        DatabaseReference myRef = database.getReference();
        DatabaseReference myref=database.getReference();
        myRef.child("Userdetail").child(uid).setValue(userData);
        myref.child(check+"chat").child(uid).setValue(new Chatusermodel(userData.name,userData.photo));
    }

    @OnClick(R.id.iview)
    void pickPhoto(){
        if(Build.VERSION.SDK_INT<Build.VERSION_CODES.M) {
            EasyImage.openChooserWithGallery(this, "Pick profile image", 0);
            //  Toast.makeText(Profile.this, "Check", Toast.LENGTH_LONG).show();
        }
        else if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.M){
            if (ContextCompat.checkSelfPermission(Signupactivity.this, android.Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(Signupactivity.this, android.Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(Signupactivity.this,
                        new String[]{android.Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.CAMERA},12345);
                return;
            }
            EasyImage.openChooserWithGallery(this, "Pick profile image", 0);
        }
    }
 @OnClick(R.id.sign_up)
    void sup(){
     InputMethodManager inputManager = (InputMethodManager)
             getSystemService(Context.INPUT_METHOD_SERVICE);

     inputManager.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(),
             InputMethodManager.HIDE_NOT_ALWAYS);
     pass=password.getEditText().getText().toString().trim();
     confrmpass=confrmpassword.getEditText().getText().toString().trim();
     if (EmailValidator.getInstance(false).isValid(email.getEditText().getText().toString().trim())) {
         email.setError(null);
         if (password.getEditText().getText().toString().trim().length()>= 6) {
             password.setError(null);
             if(confrmpass.equals(pass)) {
                 if (name.getEditText().getText().toString().trim().length() >= 4) {
                     name.setError(null);
                     if (phone.getEditText().getText().toString().trim().length() == 10) {
                         phone.setError(null);
                         if(clg.getEditText().getText().toString().trim().length()!=0){
                             clg.setError(null);
                         }else{
                             clg.setError("Enter college name");
                         }
                         attemptSignup(password.getEditText().getText().toString().trim(), email.getEditText().getText().toString().trim());
                         signup.setEnabled(false);
                     } else
                         phone.setError("Enter 10 digits");
                 } else
                     name.setError("Enter at least 4 characters");
             }else{
                 AlertDialog.Builder d = new AlertDialog.Builder(Signupactivity.this);
                 d.setMessage("Password is not matching").
                         setCancelable(true);
                 AlertDialog alert = d.create();
                 alert.setTitle("Oops...!");
                 alert.show();
             }
         } else
             password.setError("Not long enough");
     } else {
         Toast.makeText(Signupactivity.this,email.getEditText().getText().toString().trim(),Toast.LENGTH_LONG).show();
         email.setError("Invalid email");
     }
 }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (ContextCompat.checkSelfPermission(Signupactivity.this, android.Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(Signupactivity.this, android.Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(Signupactivity.this,
                    new String[]{android.Manifest.permission.READ_EXTERNAL_STORAGE, android.Manifest.permission.CAMERA},12345);
            return;
        }
        EasyImage.handleActivityResult(requestCode, resultCode, data, this, new DefaultCallback() {
            @Override
            public void onImagePickerError(Exception e, EasyImage.ImageSource source, int type) {
                //Some error handling
            }

            @Override
            public void onImagePicked(File imageFile, EasyImage.ImageSource source, int type) {
                //Handle the image
                BitmapFactory.Options options = new BitmapFactory.Options();
                options.inPreferredConfig = Bitmap.Config.ARGB_4444;
                Bitmap bitmap = BitmapFactory.decodeFile(imageFile.getAbsolutePath(), options);
                bitmap = getResizedBitmap(bitmap, 100);
                Timber.d("xxx" + bitmap.getByteCount());
                iv.setImageBitmap(bitmap);
                flag=1;
                Timber.d("flag"+flag);
                userData.photo = com.example.amit.uniconnexample.utils.Utils.encodeToBase64(bitmap, Bitmap.CompressFormat.PNG, 100);

            }
        });
    }
    public Bitmap getResizedBitmap(Bitmap image, int maxSize) {
        int width = image.getWidth();
        int height = image.getHeight();

        float bitmapRatio = (float) width / (float) height;
        if (bitmapRatio > 0) {
            width = maxSize;
            height = (int) (width / bitmapRatio);
        } else {
            height = maxSize;
            width = (int) (height * bitmapRatio);
        }
        return Bitmap.createScaledBitmap(image, width, height, true);
    }
    void attemptSignup(String pass, String id) {
      //  mProgress.setMessage("Loading....");
     //   mProgress.show();
        userData.phone = phone.getEditText().getText().toString();
        userData.name = name.getEditText().getText().toString();
        userData.email=email.getEditText().getText().toString();
        String n=userData.email;
         check=n.substring(n.indexOf("@")+1,n.lastIndexOf("."));
        userData.clg=clg.getEditText().getText().toString();
        Bitmap icon = BitmapFactory.decodeResource(getResources(),
                R.drawable.user);
        if(flag!=1) {
            userData.photo = com.example.amit.uniconnexample.utils.Utils.encodeToBase64(icon, Bitmap.CompressFormat.PNG, 100);
        }
        Timber.d(userData.photo);
        Timber.d("flag"+flag);
        phn=phone.getEditText().getText().toString().trim();
      //  pass=password.getEditText().getText().toString().trim();
       // nam=name.getEditText().getText().toString().trim();
       // mal=email.getEditText().getText().toString().trim();

        mAuth.createUserWithEmailAndPassword(id, pass)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (!task.isSuccessful()) {
                            signup.setEnabled(true);
                            Toast.makeText(Signupactivity.this, "Signup failed." + task.getException().getMessage(),
                                    Toast.LENGTH_SHORT).show();
                        }
                       /* else{
                            Toast.makeText(Signupactivity.this, "Successfully signed up", Toast.LENGTH_SHORT).show();
                            writeUserData(user.getUid());
                            mProgress.dismiss();
                            startActivity(new Intent(Signupactivity.this, MainActivity.class)
                                    .putExtra("user", userData));

                            finish();
                        }*/
                    }
                });
    }
    @Override
    public void onStart() {
        super.onStart();
        mAuth.addAuthStateListener(mAuthListener);
    }

    @Override
    public void onStop() {
        super.onStop();
        if (mAuthListener != null) {
            mAuth.removeAuthStateListener(mAuthListener);
        }
    }

}

File: app/src/main/java/com/example/amit/uniconnexample/Signupactivity.kt
package com.example.amit.uniconnexample

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.PorterDuff
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import android.view.inputmethod.InputMethodManager
import android.widget.Toast

import com.example.amit.uniconnexample.Activity.NewTabActivity
import com.example.amit.uniconnexample.MediaPicker.ChosenMediaFile
import com.example.amit.uniconnexample.MediaPicker.MediaPickerActivity
import com.example.amit.uniconnexample.Others.CommonString
import com.example.amit.uniconnexample.Others.UserData
import com.example.amit.uniconnexample.rest.RetrofitClientBuilder
import com.example.amit.uniconnexample.rest.model.ModelResponseMessage
import com.example.amit.uniconnexample.rest.model.UserDetailRequestModel
import com.example.amit.uniconnexample.utils.PrefManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.storage.FirebaseStorage
import kotlinx.android.synthetic.main.activity_signup.*

import org.apache.commons.validator.routines.EmailValidator

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException

/**
 * Created by amit on 29/10/16.
 */

class Signupactivity : AppCompatActivity() {
    internal var user: FirebaseUser? = null
    private var mProgress: ProgressDialog? = null
    private var mAuth: FirebaseAuth? = null
    lateinit var userData: UserData
    lateinit var mal: String
    lateinit var pass: String
    lateinit var confrmpass: String
    lateinit var phn: String
    lateinit var nam: String
    lateinit var clgname: String
    internal var flag = 0
    var check: String ?= ""
    var mImageUri: Uri?= null

    companion object{
        private val GALLERY_REQUEST = 1
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_signup)
        mProgress = ProgressDialog(this)
        userData = UserData()
        mAuth = FirebaseAuth.getInstance()
        val tabIconColor = ContextCompat.getColor(baseContext, R.color.md_orange_600)
        person!!.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN)
        msg!!.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN)
        lock1!!.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN)
        lock2!!.setColorFilter(tabIconColor, PorterDuff.Mode.SRC_IN)

        iview.setOnClickListener { pickPhoto() }

        sign_up.setOnClickListener{ sup() }
    }

    internal fun pickPhoto() {

        if (ContextCompat.checkSelfPermission(this, android.Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    arrayOf(android.Manifest.permission.READ_EXTERNAL_STORAGE), 12345)
            return
        }
        val intent = Intent(this, MediaPickerActivity::class.java)
        intent.putExtra(MediaPickerActivity.SINGLE_MEDIA_ALLOWED, 1)
        startActivityForResult(intent, CommonString.MEDIA_PICKER_ACTIVITY)
    }

    internal fun sup() {
        val inputManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager

        inputManager.hideSoftInputFromWindow(currentFocus!!.windowToken,
                InputMethodManager.HIDE_NOT_ALWAYS)
        pass = password!!.editText!!.text.toString().trim { it <= ' ' }
        confrmpass = confirmpassword!!.editText!!.text.toString().trim { it <= ' ' }
        if (EmailValidator.getInstance(false).isValid(email!!.editText!!.text.toString().trim { it <= ' ' })) {
            email!!.error = null
            if (password!!.editText!!.text.toString().trim { it <= ' ' }.length >= 6) {
                password!!.error = null
                if (confrmpass == pass) {
                    if (name!!.editText!!.text.toString().trim { it <= ' ' }.length >= 4) {
                        name!!.error = null

                        attemptSignup(pass, email?.editText?.text?.toString()?:"")
                        sign_up?.isEnabled = false
                    } else
                        name!!.error = "Enter at least 4 characters"
                } else {
                    val d = AlertDialog.Builder(this@Signupactivity)
                    d.setMessage("Password is not matching").setCancelable(true)
                    val alert = d.create()
                    alert.setTitle("Oops...!")
                    alert.show()
                }
            } else
                password!!.error = "Not long enough"
        } else {
            Toast.makeText(this@Signupactivity, email!!.editText!!.text.toString().trim { it <= ' ' }, Toast.LENGTH_LONG).show()
            email!!.error = "Invalid email"
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if(requestCode == CommonString.MEDIA_PICKER_ACTIVITY && data != null){
            val medias = data.getParcelableArrayListExtra<ChosenMediaFile>(CommonString.MEDIA)
            if(medias.size > 0){
                mImageUri = medias.get(0).uri
            }
        }

        mImageUri?.let {
            try {
                val bitmap = MediaStore.Images.Media.getBitmap(contentResolver, it)
                iview.setImageBitmap(bitmap)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun uploadProfile(){
        if (mImageUri != null) {
            val filepath =  FirebaseStorage.getInstance().reference.child("User_Image").child(mImageUri!!.lastPathSegment!!)
            mImageUri?.let {
                val uploadTask = filepath.putFile(it)
                uploadTask.addOnSuccessListener {
                    val downloadUrl = it.downloadUrl?.toString()?:""
                    doSingUp(downloadUrl)
                }

                uploadTask.addOnFailureListener {
                    doSingUp("")
                }

            }

        }
    }

    internal fun attemptSignup(pass: String, id: String) {
        Log.d("Sign up","attempt")
        var userId = ""
        mAuth?.createUserWithEmailAndPassword(id, pass)
                ?.addOnCompleteListener(this) { task ->
                    if (!task.isSuccessful) {
                        sign_up!!.isEnabled = true
                        Toast.makeText(this@Signupactivity, "Signup failed." + task.exception!!.message,
                                Toast.LENGTH_SHORT).show()
                    }else{
                        Log.d("Sign up","completed ${mAuth?.currentUser?.uid}")
                        mAuth?.currentUser?.getToken(false)?.addOnCompleteListener {
                            if(it.isSuccessful){
                                uploadProfile()
                            }else{
                                Log.d("sign up","auth failed")
                            }

                        }
                    }
                }
    }

    fun doSingUp(imageUrl: String){
        FirebaseAuth.getInstance()?.currentUser?.getToken(false)?.addOnCompleteListener {
            if(it.isSuccessful && it.result != null){
                val userId = FirebaseAuth.getInstance()?.currentUser?.uid?:""
                val userDetailRequestModel = UserDetailRequestModel(name?.editText?.text?.toString()?:"", imageUrl, email?.editText?.text?.toString()?:"")
                RetrofitClientBuilder(CommonString.base_url).getmNetworkRepository()?.sendUser("Bearer ${it.result?.token}", userDetailRequestModel)
                        ?.enqueue(object : Callback<ModelResponseMessage>{
                            override fun onFailure(call: Call<ModelResponseMessage>, t: Throwable) {
                                t.printStackTrace()
                                sign_up?.isEnabled = true
                            }

                            override fun onResponse(call: Call<ModelResponseMessage>, response: Response<ModelResponseMessage>) {
                                PrefManager.putString(CommonString.USER_ID, userId)
                                PrefManager.putString(CommonString.USER_NAME, name?.editText?.text?.toString()?:"")
                                PrefManager.putString(CommonString.USER_DP, imageUrl)
                                PrefManager.putString(CommonString.USER_EMAIL, email?.editText?.text?.toString()?:"")
                                val i = Intent(this@Signupactivity, NewTabActivity::class.java)
                                startActivity(i)
                                finish()
                            }
                        })
            }
        }
    }

    public override fun onStart() {
        super.onStart()
    }

    public override fun onStop() {
        super.onStop()
    }

}
